### PR TITLE
[Fix #11514] Add new `Lint/UnescapedBracketInRegexp` cop

### LINKED ISSUE #11514: Cop to catch unescaped `]` in RegExp
## Is your feature request related to a problem? Please describe.

Ruby emits a warning for the following:

```sh
ruby -c -e '/abc]123/'
-e:1: warning: regular expression has ']' without escape: /abc]123/
Syntax OK
```

## Describe the solution you'd like

It would be nice if Rubocop could catch this statically, as that integrates better into CI.

## Describe alternatives you've considered

Tools like [Deprecation Toolkit](https://github.com/Shopify/deprecation_toolkit) can be configured to catch warnings, but in the case of Deprecation Toolkit, it will only do so while running tests, whereas we encountered a case where the file had already been read by then.

### PR DESCRIPTION
**Pull request by dvandersluis**
Adds new cop to emulate the following ruby warning:

```
$ ruby -e '/abc]123/'
-e:1: warning: regular expression has ']' without escape: /abc]123/
```

Fixes #11514.

-----------------

Before submitting the PR make sure the following are checked:

* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Ran `bundle exec rake default`. It executes all tests and runs RuboCop on its own code.
* [x] Added an entry (file) to the [changelog folder](https://github.com/rubocop/rubocop/blob/master/changelog/) named `{change_type}_{change_description}.md` if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format) for details.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

**Comment by dvandersluis** [comment]
@rubocop/rubocop-core the cop parsing code that would register an offense causes ruby to print out the following warning in addition to the offense:

```
$ rubocop test.rb
Inspecting 1 file
/Users/daniel/.rbenv/versions/3.3.5/lib/ruby/gems/3.3.0/gems/parser-3.3.5.0/lib/parser/builders/default.rb:2266: warning: regular expression has ']' without escape: /abc]123/
W

Offenses:

test.rb:3:5: W: [Correctable] Lint/UnescapedBracketInRegexp: Regular expression has ] without escape.
/abc]123/
    ^
```

Any idea how to suppress it? I tried wrapping `on_regexp` with `RuboCop::Util.silence_warnings` but it didn't help. Even having an empty `on_regexp` method triggers the warning from parser.

**Comment by Earlopain** [comment]
This isn't the only such code that would print such things. `FloatOutOfRange` suffers from the same thing, among a few other cops. These happen during parsing. I openend https://github.com/whitequark/parser/pull/1013 to supress these but there was some concern and I ended up not needing the PR myself anymore.

In theory you can wrap creation of the ProcessedSource in `rubocop-ast` to silence these warnings but I find that to be too broad.

**Comment by dvandersluis** [comment]
@owst @bbatsov I missed that this was previously tackled in #12805 which was recently closed. 

Although there was previous discussion about this being too specific, I think it does make sense given this is a Lint cop emulating a specific ruby warning. If there are other similar warnings they could be handled too but I didn't find any. 

I also handled the unescaped bracket being the first character, as discussed in that cop.

**Comment by bbatsov** [comment]
That's fair. I had kind of missed the point that this was emulating an existing Ruby warning, so being specific is fine here.

File: config/default.yml [hunk 1]
```diff
@@ -2457,6 +2457,11 @@ Lint/UnderscorePrefixedVariableName:
   VersionAdded: '0.21'
   AllowKeywordBlockArguments: false
 
+Lint/UnescapedBracketInRegexp:
+  Description: 'Checks for unescaped literal `[` in Regexp.'
```
Comment (bquorning) [inline config/default.yml:2461]: I think you mean `]` here.

File: lib/rubocop/cop/lint/unescaped_bracket_in_regexp.rb [hunk 2]
```diff
@@ -0,0 +1,88 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Lint
+      # Checks for Regexpes (both literals and via `Regexp.new` / `Regexp.compile`)
+      # that contain unescaped `[` characters.
```
Comment (bquorning) [inline lib/rubocop/cop/lint/unescaped_bracket_in_regexp.rb:7]: I think you mean `]` here too.

**Review by bquorning — COMMENTED** [review]
Instead of having a RuboCop cop for this, shouldn’t we just recommend that people use https://github.com/rubocop/rubocop-rubycw ?

File: config/default.yml (hunk 1 above)
Comment (dvandersluis) [reply config/default.yml:2461]: 🤦 thank you!

**Comment by dvandersluis** [comment]
@bquorning honestly I didn't know that existed, but I think it's still ok because this cop also autocorrects. In any case, there are quite a few existing cops that mirror ruby warnings.

**Comment by bbatsov** [comment]
A bit of context for both of you - originally something very similar to `rubocop-rubycw` was part of RuboCop, but I've decided to emulate the built-in warnings one by one as cops, so that the warnings can be improved in one or another. (e.g. have more informative messages) I didn't want them to be carbon copy replica of the default Ruby behavior and we always have some leeway to do something a bit different/better than whatever's built-in.

**Comment by bquorning** [comment]
Thank you for the extra context @bbatsov.

I was thinking for a minute that perhaps these cops solving `ruby -cw` problems belong better in `rubocop-rubycw`, but upon further thought no. In my (current) opinion, that gem is perfect the way it is, and the more specialized cops with autocorrection and whatnot belong just fine in our `Lint` namespace.

**Comment by Earlopain** [comment]
`rubocop-rubycw` is nice but relies on mri-specific APIs that are unavailable on jruby and others. For those, it shells out once per source file which probably isn't the nicest of experience.

**Pull request merged**